File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.java
class Solution {
    public long[] resultArray(int[] nums, int k) {
        long[] answer = new long[k];
        long[] dp = new long[k];

        for (int num : nums) {
            int value = num % k;
            long[] next = new long[k];

            // start a new subarray with nums[i]
            next[value]++;

            // extend evry subarray ending at the previous index
            for (int r = 0; r < k; r++) {
                if (dp[r] > 0) {
                    int newRemainder = (r * value) % k;
                    next[newRemainder] += dp[r];
                }
            }

            // add all subarrays ending at this index
            for (int r = 0; r < k; r++)
                answer[r] += next[r];
            
            dp = next;
        }

        return answer;
    }
}

File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.py
class Solution:
    def resultArray(self, nums: list[int], k: int) -> list[int]:
        answer = [0] * k
        dp = [0] * k

        for num in nums:
            value = num % k
            next_dp = [0] * k

            # start a new subarray with nums[i]
            next_dp[value] += 1

            # extend every subarray ending at the previous index
            for r in range(k):
                if dp[r] > 0:
                    new_remainder = (r * value) % k
                    next_dp[new_remainder] += dp[r]

            # add all subarrays ending at this index
            for r in range(k):
                answer[r] += next_dp[r]
                
            dp = next_dp

        return answer
